python/plugins/processing/algs/otb/OtbAlgorithm.py

```diff
@@ -0,0 +1,246 @@
+                    if name in ["elev.dem.geoid", "elev.geoid"]:
+                        param.setDefaultValue(OtbUtils.geoidFile())
+
+                    self.addParameter(param)
+                    #parameter is added now and we must move to next line
+                    line = lines.readline().strip('\n').strip()
+
+        except BaseException as e:
+            import traceback
+            errmsg = "Could not open OTB algorithm from file: \n" + self._descriptionfile + "\nline=" + line + "\nError:\n" + traceback.format_exc()
+            QgsMessageLog.logMessage(self.tr(errmsg), self.tr('Processing'), Qgis.Critical)
+            raise e
+
+    def preprocessParameters(self, parameters):
+        valid_params = {}
+        for k, v in parameters.items():
+            param = self.parameterDefinition(k)
+            #If parameterDefinition(k) return None, this is considered a invalid parameter.
+            #just continue for loop
+            if param is None:
+                continue
+
+            #if name of parameter is 'pixtype',
+            #it is considered valid if it has value other than float
+            if k == 'outputpixeltype' and self.pixelTypes[int(v)] == 'float':
+                continue
+            # Any other valid parameters have:
+            #- empty or no metadata
+            #- metadata without a 'group_key'
+            #- metadata with 'group_key' and 'group_key' is not in list of parameters. see ParameterGroup in OTB
+            #- metadata with 'group_key' and 'group_key' is a valid parameter and it's value == metadata()['group_value']
+            if 'group_key' in param.metadata() and not param.metadata()['group_key'] in parameters:
+                valid_params[k] = v
+            elif not 'group_key' in param.metadata() or parameters[param.metadata()['group_key']] == param.metadata()['group_value']:
+                valid_params[k] = v
+
+        return valid_params
+
+    def otbParameterValue(self, v):
+        if isinstance(v, QgsMapLayer):
+            return v.source()
+        elif isinstance(v, QgsProcessingOutputLayerDefinition):
+            return v.sink.staticValue()
+        else:
+            return str(v)
+
+    def outputParameterName(self):
+        with open(self._descriptionfile) as df:
+            first_line = df.readline().strip()
+            tokens = first_line.split("|")
+            #params = [t if str(t) != str(None) else None for t in tokens[1:]]
+            if len(tokens) == 2:
+                return tokens[1]
+            else:
+                return ''
+
+    def processAlgorithm(self, parameters, context, feedback):
+        output_key = self.outputParameterName()
+        otb_cli_file = OtbUtils.cliPath()
+        command = '"{}" {} {}'.format(otb_cli_file, self.name(), OtbUtils.appFolder())
+        for k, v in parameters.items():
+            if k == 'outputpixeltype' or not v:
+                continue
+
+            if 'epsg' in k and v.startswith('EPSG:'):
```

Comment:
otb only uses QgsCoordinateReferenceSystem for those keys.  attached below a grep of epsg key used in otb descriptors. So I think it is missing an instance check on QgsCoordinateReferenceSystem. 

```
[~/projects/qgis/otb/build/share/otb/description]
$ grep -rs epsg.code .
./GenerateRPCSensorModel.txt:QgsProcessingParameterCrs|map.epsg.code|EPSG Code|ProjectCrs|True
./ConvertCartoToGeoPoint.txt:QgsProcessingParameterCrs|mapproj.epsg.code|EPSG Code|ProjectCrs|True
./OrthoRectification.txt:QgsProcessingParameterCrs|map.epsg.code|EPSG Code|ProjectCrs|True
./VectorDataReprojection.txt:QgsProcessingParameterNumber|out.proj.user.map.epsg.code|EPSG Code|QgsProcessingParameterNumber.Integer|4326|True
./RefineSensorModel.txt:QgsProcessingParameterCrs|map.epsg.code|EPSG Code|ProjectCrs|True
./PixelValue.txt:QgsProcessingParameterCrs|mode.epsg.code|EPSG code|ProjectCrs|True
./GeneratePlyFile.txt:QgsProcessingParameterCrs|map.epsg.code|EPSG Code|ProjectCrs|True
./StereoFramework.txt:QgsProcessingParameterCrs|map.epsg.code|EPSG Code|ProjectCrs|True
```

Reply:
Actually the algorithms need to be able to handle a very wide variety of input types -- e.g. for CRS parameters valid inputs include QgsCoordinateReferenceSystem objects, authid strings ("epsg:3111"), proj strings, etc. The only way to gaurantee that your algorithm is handling everything correctly is to use the built-in QgsProcessingAlgorithm::parameterAs* methods. I'd expect to see a large block like https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/saga/SagaAlgorithm.py#L254 which handles translation of the evaluated parameter values into the formats required by OTB.

Reply:
okay. I think otb is much less of mess compared to SAGA processing starting with no need to export to specific raster/vector formats.

* For boolean values: True with lower and upper case 't' are accepted by OTB. It always takes into lower case and then set parameter for otb algorithm.
* For numeric values: otb can take float/double for int but only int part is taken. So it doesn't matter to fixup values
* For QgsProcessingParameterString, QgsProcessingParameterFile: it always needs  full path of file or string value.  This is already handled by gui/wrappers.py 
* For QgsProcessingParameterRasterLayer: there is already a method to return file name instead if instance is QgsMapLayer. so that also works for Vectors
* For QgsProcessingParameterExtent, QgsProcessingParameterMatrix, QgsProcessingParameterFeatureSource:  otb is not using them at this point. 
* For QgsProcessingParameterEnum: only enum used is for outputpixeltype which is an advanced parameter for all applications that holds pixeltypes. It doesn't matter it returns the string as is for combobox. It is handled specially in processAlgorithm.

Do you have a use case where the current code is breaking?. I will try to push an update of processAlgorithm for Otb.

Reply:
pushed a change to use parameterAs method. I will have to add a test on one more algorithm that uses Crs parameter. Will push that if new changes to OtbAlgorithm.py is okay.

Reply:
@rkanavath I realise that OTB situation is easier than SAGA here, but you still need to use the parameterAs* methods, regardless of the parameter values that the OTB algorithm dialog and the wrappers themselves returns. This is because Processing algorithms are also called from scripts and the command line, and using output values from other algorithms when inside a larger model -- so something like the crs parameter HAS to handle QgsCoordinateReferenceSystem values, proj strings, epsg codes, etc. The only way to guarantee that your algorithm will work everywhere as expected is to use the base algorithm parameterAs* methods.

Reply:
Thanks!